x/ccv/provider/keeper/misbehaviour.go

```diff
@@ -0,0 +1,165 @@
+	// w.r.t to the last trusted consensus it entails that the infraction age
+	// isn't too old. see ibc-go/modules/light-clients/07-tendermint/types/misbehaviour_handle.go
+
+	// Get Byzantine validators from the conflicting headers
+	// Note that it returns an error if the misbehaviour doesn't correspond to an equivocation
+	byzantineValidators, err := k.GetByzantineValidators(ctx, misbehaviour)
+	if err != nil {
+		return err
+	}
+
+	// jail and tombstone the Byzantine validators
+	for _, v := range byzantineValidators {
+		// convert consumer consensus address
+		consuAddr := sdk.ConsAddress(v.Address.Bytes())
+		provAddr := k.GetProviderAddrFromConsumerAddr(ctx, misbehaviour.Header1.Header.ChainID, types.NewConsumerConsAddress(consuAddr))
+		k.stakingKeeper.ValidatorByConsAddr(ctx, consuAddr)
+		val, ok := k.stakingKeeper.GetValidatorByConsAddr(ctx, provAddr.Address)
+
+		if !ok || val.IsUnbonded() {
+			logger.Error("validator not found or is unbonded", provAddr.String())
+			continue
+		}
+
+		// jail validator if not already
+		if !val.IsJailed() {
+			k.stakingKeeper.Jail(ctx, provAddr.Address)
+		}
+
+		// tombstone validator if not already
+		if !k.slashingKeeper.IsTombstoned(ctx, provAddr.Address) {
+			k.slashingKeeper.Tombstone(ctx, provAddr.Address)
+		}
+
+		// update jail time to end after double sign jail duration
+		k.slashingKeeper.JailUntil(ctx, provAddr.Address, evidencetypes.DoubleSignJailEndTime)
+	}
+
+	logger.Info(
+		"confirmed equivocation",
+		"byzantine validators", byzantineValidators,
+	)
+
+	return nil
+}
+
+// GetByzantineValidators returns the Byzantine validators from a given misbehaviour
+// with the condition that it corresponds to an equivocation light client attack
+func (k Keeper) GetByzantineValidators(ctx sdk.Context, misbehaviour ibctmtypes.Misbehaviour) ([]*tmtypes.Validator, error) {
+	// construct the trusted and conflicted light blocks
+	header1, err := headerToLightBlock(*misbehaviour.Header1)
+	if err != nil {
+		return nil, err
+	}
+	header2, err := headerToLightBlock(*misbehaviour.Header2)
+	if err != nil {
+		return nil, err
+	}
+
+	var validators []*tmtypes.Validator
+
+	// compare the signatures of the headers
+	// and return the intersection of validators who signed both
+
+	// create a map with the validators' address that signed header1
+	header1Signers := map[string]struct{}{}
+	for _, sign := range header1.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		header1Signers[sign.ValidatorAddress.String()] = struct{}{}
+	}
+
+	// iterate over the header2 signers
+	// and check if they s signed header1
+	for _, sign := range header2.Commit.Signatures {
+		if sign.Absent() {
+			continue
+		}
+		if _, ok := header1Signers[sign.ValidatorAddress.String()]; ok {
+			_, val := header1.ValidatorSet.GetByAddress(sign.ValidatorAddress)
+			validators = append(validators, val)
+		}
+	}
+
+	sort.Sort(tmtypes.ValidatorsByVotingPower(validators))
```

Comment:
Why is this sort needed?

Reply:
Good point. Comet returns the byzantine validators ordered by val powers but not sure if it's required here.
Any idea @mpoke ?